xclim/cli.py

```diff
@@ -0,0 +1,239 @@
+
+
+def _get_indicator(indname):
+    if "." in indname:
+        modname, indname = indname.split(".")
+    else:
+        raise click.BadArgumentUsage(
+            f"Indicator name must include the module name (ex: atmos.tg_mean) (got {indname})"
+        )
+
+    if modname not in xcmodules or indname not in xcmodules[modname].__dict__:
+        raise click.BadArgumentUsage(
+            f"Indicator '{indname}' or module '{modname}' not found in xclim."
+        )
+
+    return xcmodules[modname].__dict__[indname]
+
+
+def _get_input(ctx):
+    arg = ctx.obj["input"]
+    if isinstance(arg, xr.Dataset):
+        return arg
+    elif isinstance(arg, tuple) or "*" in arg:
+        ctx.obj["xr_kwargs"].setdefault("combine", "by_coords")
+        ds = xr.open_mfdataset(arg, **ctx.obj["xr_kwargs"])
+    else:
+        ctx.obj["xr_kwargs"].pop("combine", None)
+        ds = xr.open_dataset(arg, **ctx.obj["xr_kwargs"])
+    ctx.obj["input"] = ds
+    return ds
+
+
+def _get_output(ctx):
+    if "ds_out" not in ctx.obj:
+        dsin = _get_input(ctx)
+        ctx.obj["ds_out"] = dsin.drop_vars(dsin.data_vars)
+    return ctx.obj["ds_out"]
+
+
+def _process_indicator(indicator, ctx, **params):
+    click.echo(
+        f"Processing : {indicator.format({'long_name': indicator.long_name}, params)['long_name']}"
+    )
+    dsin = _get_input(ctx)
+    dsout = _get_output(ctx)
+
+    for key, val in params.items():
+        if (
+            indicator._sig.parameters[key].default is inspect._empty
+        ):  # a Dataset is expected
+            if key == "tas" and val is None and key not in dsin.data_vars:
+                # Special case for tas.
+                params["tas"] = atmos.tg(dsin.tasmin, dsin.tasmax)
+            else:
+                params[key] = dsin[
+                    val or key
+                ]  # Either a variable name was given or the key is the name
+
+    var = indicator(**params)
+    dsout = dsout.assign({var.name: var})
+    ctx.obj["ds_out"] = dsout
+
+
+def _create_command(name):
+    indicator = _get_indicator(name)
+    params = []
+    for name, param in indicator._sig.parameters.items():
+        params.append(
+            click.Option(
+                param_decls=["--" + name],
+                default=None if param.default is inspect._empty else param.default,
+                show_default=True,
+                help=indicator._parameters_doc.get(name),
+            )
+        )
+
+    @click.pass_context
+    def _process(ctx, **kwargs):
+        return _process_indicator(indicator, ctx, **kwargs)
+
+    return click.Command(
+        name,
+        callback=_process,
+        params=params,
+        help=indicator.abstract,
+        short_help=indicator.long_name,
+    )
+
+
+@click.command(short_help="List indicators.")
+@click.argument("module", required=False, nargs=-1)
+@click.option(
+    "-i", "--info", is_flag=True, help="Prints more details for each indicator."
+)
+def list(module, info):
```

Comment:
Codacity is raising errors over this redeclaration. If this is the proper way of using click, maybe adding a one exemption to the code linter here would be good.

Reply:
Or we could choose another name, like "list-indices", or "indices". Or use the `get_command` to return a properly-named function when the uses enters "list".